pkg/operator/controllers/imageconfig/image_controller.go

```diff
@@ -0,0 +1,154 @@
+)
+
+const imageConfigResource = "cluster"
+
+type Reconciler struct {
+	log       *logrus.Entry
+	arocli    aroclient.Interface
+	configcli configclient.Interface
+}
+
+func NewReconciler(log *logrus.Entry, arocli aroclient.Interface, configcli configclient.Interface) *Reconciler {
+	return &Reconciler{
+		log:       log,
+		arocli:    arocli,
+		configcli: configcli,
+	}
+}
+
+// watches the ARO object for changes and reconciles image.config.openshift.io/cluster object.
+// - If blockedRegistries is not nil, makes sure required registries are not added
+// - If AllowedRegistries is not nil, makes sure required registries are added
+// - Fails fast if both are not nil, unsupported
+func (r *Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	// Get cluster
+	instance, err := r.arocli.AroV1alpha1().Clusters().Get(ctx, request.Name, metav1.GetOptions{})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Feature flag
+	if !instance.Spec.Features.ReconcileImageConfig {
+		return reconcile.Result{}, nil
+	}
+
+	// Check for cloud type
+	requiredRegistries, err := getCloudAwareRegistries(instance)
+	if err != nil {
+		// Not returning error as it will requeue again
+		return reconcile.Result{Requeue: false}, nil
+	}
+
+	// Get image.config yaml
+	imageconfig, err := r.configcli.ConfigV1().Images().Get(ctx, request.Name, metav1.GetOptions{})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Fail fast if both are not nil
+	if imageconfig.Spec.RegistrySources.AllowedRegistries != nil && imageconfig.Spec.RegistrySources.BlockedRegistries != nil {
+		err := errors.New("both AllowedRegistries and BlockedRegistries are present")
+		return reconcile.Result{}, err
+	}
+
+	// Append to allowed registries
+	if imageconfig.Spec.RegistrySources.AllowedRegistries != nil {
+		removeDuplicateRegistries := func(item string) bool {
+			for _, v := range requiredRegistries {
+				if strings.EqualFold(item, v) {
+					return false
+				}
+			}
+			return true
+		}
+		imageconfig.Spec.RegistrySources.AllowedRegistries = filterSliceInPlace(imageconfig.Spec.RegistrySources.AllowedRegistries, removeDuplicateRegistries)
+		imageconfig.Spec.RegistrySources.AllowedRegistries = append(imageconfig.Spec.RegistrySources.AllowedRegistries, requiredRegistries...)
+	}
+
+	// Remove from blocked registries
+	if imageconfig.Spec.RegistrySources.BlockedRegistries != nil {
+		removeRequiredRegistries := func(item string) bool {
+			for _, v := range requiredRegistries {
+				if strings.EqualFold(item, v) {
+					return false
+				}
+			}
+			return true
+		}
+		imageconfig.Spec.RegistrySources.BlockedRegistries = filterSliceInPlace(imageconfig.Spec.RegistrySources.BlockedRegistries, removeRequiredRegistries)
+	}
+
+	// Update image config registry
+	_, err = r.configcli.ConfigV1().Images().Update(ctx, imageconfig, metav1.UpdateOptions{})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	return reconcile.Result{}, nil
```

Comment:
And another minor suggestion: we can simplify this by always returning `err` which will either be `nil` or non-`nil`.

```suggestion
	_, err = r.configcli.ConfigV1().Images().Update(ctx, imageconfig, metav1.UpdateOptions{})
	return reconcile.Result{}, err
```